text/0000-async-stream.md

```diff
@@ -0,0 +1,792 @@
+
+[iterator]: https://doc.rust-lang.org/std/iter/trait.Iterator.html
+[`Future`]: https://doc.rust-lang.org/std/future/trait.Future.html
+[`Poll::Pending`]: https://doc.rust-lang.org/std/task/enum.Poll.html#variant.Pending
+
+```rust
+// Defined in std::stream module
+pub trait Stream {
+    // Core items:
+    type Item;
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>>;
+    
+    // Optional optimization hint, just like with iterators:
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, None)
+    }
+}
+```
+
+The arguments to `poll_next` match that of the [`Future::poll`] method:
+
+* The self must be a pinned reference, ensuring both unique access to
+  the stream and that the stream value itself will not move. Pinning
+  allows the stream to save pointers into itself when it suspends,
+  which will be required to support generator syntax at some point.
+* The [context] `cx` defines details of the current task. In particular,
+  it gives access to the [`Waker`] for the task, which will allow the
+  task to be re-awoken once data is ready.
+
+[`Future::poll`]: https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll
+[pinned]: https://doc.rust-lang.org/std/pin/struct.Pin.html
+[context]: https://doc.rust-lang.org/std/task/struct.Context.html
+[`Waker`]: https://doc.rust-lang.org/std/task/struct.Waker.html
+
+### Usage
+
+A user could create a stream as follows (Example taken from @yoshuawuyt's [implementation pull request](https://github.com/rust-lang/rust/pull/79023)).
+
+Creating a stream involves two steps: creating a `struct` to
+ hold the stream's state, and then implementing `Stream` for that
+ `struct`.
+
+ Let's make a stream named `Counter` which counts from `1` to `5`:
+
+```rust
+#![feature(async_stream)]
+# use core::stream::Stream;
+# use core::task::{Context, Poll};
+# use core::pin::Pin;
+
+// First, the struct:
+
+/// A stream which counts from one to five
+struct Counter {
+    count: usize,
+}
+
+// we want our count to start at one, so let's add a new() method to help.
+// This isn't strictly necessary, but is convenient. Note that we start
+// `count` at zero, we'll see why in `poll_next()`'s implementation below.
+impl Counter {
+    fn new() -> Counter {
+        Counter { count: 0 }
+    }
+}
+
+// Then, we implement `Stream` for our `Counter`:
+
+impl Stream for Counter {
+    // we will be counting with usize
+    type Item = usize;
+
+    // poll_next() is the only required method
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Increment our count. This is why we started at zero.
+        self.count += 1;
+
+        // Check to see if we've finished counting or not.
+        if self.count < 6 {
+            Poll::Ready(Some(self.count))
+        } else {
+            Poll::Ready(None)
+        }
+    }
+}
+```
+
+## Initial impls
```

Comment:
I found myself wanting to see the contents of the _More Usage Examples_ section here first after we defined our `Counter` to see what it's like to interact with on the consuming side.

Reply:
Hi @KodrAus  - I intentionally did not include the consuming side here since the `next` method has been moved to the "Future Possibilities" section. I felt including that here would introduce more confusion about what is and is not covered by this RFC.